pkg/comm/resolve_comm.go

```diff
@@ -0,0 +1,89 @@
+package comm
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"golang.org/x/sys/unix"
+)
+
+const unknownComm = "unknown"
+
+type CommResolver struct {
+	cacheExist    map[int]string
+	cacheNotExist map[int]bool
+}
+
+func NewCommResolver() *CommResolver {
+	return &CommResolver{
+		cacheExist:    map[int]string{},
+		cacheNotExist: map[int]bool{},
+	}
+}
+
+func (r *CommResolver) ResolveComm(pid int) (string, error) {
+	if comm, ok := r.cacheExist[pid]; ok {
+		return comm, nil
+	}
+	if _, ok := r.cacheNotExist[pid]; ok {
+		return unknownComm, fmt.Errorf("process not running")
+	}
+
+	comm, err := readCommandFromProcFs(pid)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// skip process that is not running
+			r.cacheNotExist[pid] = true
+			return unknownComm, fmt.Errorf("process not running: %w", err)
+		}
+	}
+	r.cacheExist[pid] = comm
```

Comment:
when do we invalidate both caches?

Reply:
fixed. we now listen for `sched_process_free` events and evict stale PIDs from both caches every sample period.